### PR TITLE
Feat/dv 136/create endpoint to confirm goods delivery

### PR DESCRIPTION
**Pull request by Grasmus**

### CODE REVIEW

**Review by AndriiChepeliuk — APPROVED** [review]

File: API/Controllers/OfferController.cs [hunk 1]
```diff
@@ -70,5 +70,16 @@ public async Task<IActionResult> DeleteAsync([FromQuery] OfferIdDTO offerIdDTO)
 
             return Ok();
         }
+
+        [HttpPost("confirm-goods-transfer")]
+        [AuthorizeByRole(IdentityRoleNames.User)]
+        public async Task<IActionResult> ConfirmGoodsTransferAsync([FromBody] ConfirmGoodsTransferDTO confirmGoodsTransferDTO)
```
Comment (marynakernychna) [inline API/Controllers/OfferController.cs]: Move the parameter to a new line

File: Core/DTO/OfferDTO/ConfirmGoodsTransferDTO.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Core.DTO.OfferDTO
+{
+    public class ConfirmGoodsTransferDTO
+    {
+        public OfferIdDTO OfferIdDTO { get; set; }
```
Comment (marynakernychna) [inline Core/DTO/OfferDTO/ConfirmGoodsTransferDTO.cs]: Why don't you just write a field?

File: Core/Interfaces/CustomService/IOfferService.cs [hunk 3]
```diff
@@ -12,5 +12,6 @@ public interface IOfferService
         Task<PaginatedList<OfferPreviewDTO>> GetUsersOffersAsync(string userId, PaginationFilterDTO paginationFilter);
         Task<PaginatedList<OfferCreateTripDTO>> GetNearRouteAsync(PaginationFilterDTO paginationFilter, int tripId);
         Task DeleteAsync(OfferIdDTO offerIdDTO, string userId);
+        Task ConfirmGoodsTransferAsync(ConfirmGoodsTransferDTO confirmGoodTransferDTO, string userId);
```
Comment (marynakernychna) [inline Core/Interfaces/CustomService/IOfferService.cs]: Move the parameter to a new line & also in the whole file

File: Core/Resources/ErrorMessages.resx [hunk 4]
```diff
@@ -249,7 +249,14 @@
...
+    <value>You can't confirm goods delivery on trip you didn't create</value>
+  </data>
+  <data name="ConfirmGoodsDeliveryWrongRoleName" xml:space="preserve">
+    <value>Such role does not exist: </value>
+  </data>
   <data name="PointsDoNotMatch" xml:space="preserve">
     <value>Points do not match with points from trip</value>
   </data>
 </root>
+
```
Comment (marynakernychna) [inline Core/Resources/ErrorMessages.resx]: Remove the extra blank line

File: Core/Services/OfferService.cs [hunk 5]
```diff
@@ -155,5 +157,83 @@ public async Task DeleteAsync(OfferIdDTO offerIdDTO, string userId)
...
+
+        public async Task ConfirmGoodsTransferAsync(
+            ConfirmGoodsTransferDTO сonfirmGoodsTransferDTO, 
+            string userId)
+        {
+            if (сonfirmGoodsTransferDTO.TripRole.ToUpper() == Constants.TripRoles.Driver)
+            {
+                var offer = await _offerRepository
+                    .GetBySpecAsync(new OfferSpecification
+                    .GetByIdWithActiveTrip(сonfirmGoodsTransferDTO.OfferIdDTO.OfferId));
```
Comment (marynakernychna) [inline Core/Services/OfferService.cs]: Add a tab before calling the specification to better understand the level of nesting. Do this for the entire file as well

File: Core/Services/OfferService.cs [hunk 6]
```diff
@@ -155,5 +157,83 @@ public async Task DeleteAsync(OfferIdDTO offerIdDTO, string userId)
             await _offerRepository.DeleteAsync(offer);
             await _pointRepository.DeleteAsync(offer.Point);
         }
+
+        public async Task ConfirmGoodsTransferAsync(
+            ConfirmGoodsTransferDTO сonfirmGoodsTransferDTO, 
+            string userId)
+        {
+            if (сonfirmGoodsTransferDTO.TripRole.ToUpper() == Constants.TripRoles.Driver)
```
Comment (marynakernychna) [inline Core/Services/OfferService.cs]: You can use "switch", it'll work faster

File: Core/Specifications/OfferSpecification.cs [hunk 7]
```diff
@@ -112,5 +112,27 @@ public GetOpenByIdAndUserIdWithoutTrip(int offerId, string userId)
...
         }
+
+        public class GetByIdWithActiveTrip : Specification<Offer>, ISingleResultSpecification<Offer>
+        {
+            public GetByIdWithActiveTrip(int offerId)
+            {
+                Query
+                    .Where(o => o.Id == offerId 
+                    && o.Trip.IsActive 
+                    && !o.IsAnsweredByDriver)
```
Comment (marynakernychna) [inline Core/Specifications/OfferSpecification.cs]: Add a tab before new conditions to better understand the level of nesting. Do this for the entire file as well

File: Core/Services/OfferService.cs [hunk 8]
```diff
@@ -155,5 +157,83 @@ public async Task DeleteAsync(OfferIdDTO offerIdDTO, string userId)
...
+            {
+                var offer = await _offerRepository
+                    .GetBySpecAsync(new OfferSpecification
+                    .GetByIdWithActiveTrip(сonfirmGoodsTransferDTO.OfferIdDTO.OfferId));
+
+                ExceptionMethods.OfferNullCheck(offer);
+
+                if (offer.Trip.TripCreatorId != userId)
```
Comment (marynakernychna) [inline Core/Services/OfferService.cs]: Why don't you check this right away in the specification?

File: Core/Specifications/OfferSpecification.cs [hunk 9]
```diff
@@ -112,5 +112,27 @@ public GetOpenByIdAndUserIdWithoutTrip(int offerId, string userId)
                     .Include(o => o.Point);
             }
         }
+
+        public class GetByIdWithActiveTrip : Specification<Offer>, ISingleResultSpecification<Offer>
```
Comment (marynakernychna) [inline Core/Specifications/OfferSpecification.cs]: Move the inheritance from the second interface to a new line

File: Core/Specifications/TripSpecification.cs [hunk 10]
```diff
@@ -88,6 +88,17 @@ public GetActiveByUserId(string userId)
                 Query.Where(t => t.TripCreatorId == userId && t.IsActive);
             }
         }
+
+        internal class GetActiveByUserIdWithOffers : Specification<Trip>, ISingleResultSpecification<Trip>
```
Comment (marynakernychna) [inline Core/Specifications/TripSpecification.cs:99]: Move the imitation from the second interface to a new line

**Review by marynakernychna — CHANGES_REQUESTED** [review]
If the driver presses the button before any of the senders or receivers, then we never run out of trip

**Review by tonia4ik-s — APPROVED** [review]

**Review by marynakernychna — APPROVED** [review]

**Pull request merged**